components/server/src/azure-devops/azure-context-parser.ts

```diff
@@ -0,0 +1,313 @@
+/**
+ * Copyright (c) 2024 Gitpod GmbH. All rights reserved.
+ * Licensed under the GNU Affero General Public License (AGPL).
+ * See License.AGPL.txt in the project root for license information.
+ */
+
+import { injectable, inject } from "inversify";
+
+import { AzureDevOpsApi } from "./azure-api";
+import { IContextParser, AbstractContextParser, URLParts } from "../workspace/context-parser";
+import { AzureDevOpsTokenHelper } from "./azure-token-helper";
+import { NavigatorContext, PullRequestContext, User, WorkspaceContext } from "@gitpod/gitpod-protocol";
+import { TraceContext } from "@gitpod/gitpod-protocol/lib/util/tracing";
+import { log } from "@gitpod/gitpod-protocol/lib/util/logging";
+import { NotFoundError, UnauthorizedError } from "../errors";
+import { getProjectAndRepoName, normalizeBranchName, toBranch, toRepository } from "./azure-converter";
+import { ApplicationError, ErrorCodes } from "@gitpod/gitpod-protocol/lib/messaging/error";
+import { AuthProviderParams } from "../auth/auth-provider";
+
+@injectable()
+export class AzureDevOpsContextParser extends AbstractContextParser implements IContextParser {
+    @inject(AzureDevOpsApi) protected readonly azureDevOpsApi: AzureDevOpsApi;
+    @inject(AzureDevOpsTokenHelper) protected readonly tokenHelper: AzureDevOpsTokenHelper;
+    @inject(AuthProviderParams) readonly config: AuthProviderParams;
+
+    public async handle(ctx: TraceContext, user: User, contextUrl: string): Promise<WorkspaceContext> {
+        const span = TraceContext.startSpan("AzureDevOpsContextParser", ctx);
+        span.setTag("contextUrl", contextUrl);
+
+        try {
+            const {
+                host,
+                owner: azOrganization,
+                repoName: projectAndRepo,
+                moreSegments,
+                searchParams,
+            } = await this.parseURL(user, contextUrl);
+            const [azProject, repoName] = getProjectAndRepoName(projectAndRepo);
+            if (moreSegments.length > 0) {
+                switch (moreSegments[0]) {
+                    case "pullrequest": {
+                        return await this.handlePullRequestContext(
+                            user,
+                            host,
+                            azOrganization,
+                            azProject,
+                            repoName,
+                            parseInt(moreSegments[1]),
+                        );
+                    }
+                    case "commit": {
+                        return await this.handleCommitContext(
+                            user,
+                            host,
+                            azOrganization,
+                            azProject,
+                            repoName,
+                            moreSegments[1],
+                        );
+                    }
+                }
+            }
```

Comment:
@mustard-mh are there plans to support issue contexts? I am totally fine with it if not, but want to know so that we can align the implementation in https://github.com/gitpod-io/browser-extension/pull/157.

Reply:
@filiptronicek Good call

Btw, there's possible urls of issue (`id=3`) I found
- https://dev.azure.com/services-azure/open-to-edit-project/_workitems/edit/3
- https://dev.azure.com/services-azure/open-to-edit-project/_boards/board/t/open-to-edit-project%20Team/Issues/?workitem=3
- https://dev.azure.com/services-azure/open-to-edit-project/_backlogs/backlog/open-to-edit-project%20Team/Issues/?workitem=3

Reply:
@filiptronicek  I just realize that issues don't belong to any repo, so seems we can't make it?

Reply:
@mustard-mh let's not implement it from the get-go and wait for someone to complain. I don't see the issue opening as an integral Gitpod feature.